components/builder-events/src/event.rs

```diff
@@ -53,3 +53,22 @@ impl BuilderEvent {
 
     pub fn routing_key(self) -> Option<String> { self.routing_key }
 }
+
+#[macro_export]
+macro_rules! publish_event {
+    ($state:expr, $type:expr, $json:expr) => {
+        publish_event!($state, $type, $json, None)
+    };
+    ($state:expr, $type:expr, $json:expr, $routing_key:expr) => {
+        if let Some(bus) = $state.eventbus.as_ref() {
+            let event_type = $type;
+            match BuilderEvent::new(event_type, $routing_key, $json) {
+                Ok(event) => bus.inner.send(event).await,
+                Err(err) => {
+                    debug!("Failed to create EventBus event for {}: {}",
+                           event_type, err)
+                }
+            }
+        }
+    };
+}
```

Comment:
I think this could just be a plain old function call, rather than a macro.

Reply:
The reason I chose a macro is to simplify the calling sites when `routing_key` is not needed and is the `None` option variant.  In that case, the macro passes the `None` for you.

```macro_rules! publish_event {
    ($opt_bus:expr, $event_type:expr, $data:expr) => {
        publish_event!($opt_bus, $event_type, $data, with_routing_key: None <--- HERE )
    };
    ($opt_bus:expr, $event_type:expr, $data:expr,with_routing_key: $routing_key:expr) => {
        if let Some(bus) = $opt_bus.as_ref() {
            match BuilderEvent::new($event_type, $routing_key, json!($data)) {
                Ok(event) => bus.send(event).await,
                Err(err) => {
                    debug!("Failed to create EventBus event for {}: {}",
                           $event_type, err)
                }
            }
        }
    };
}
```

Reply:
Ah... I missed the `with_routing_key` bit in the match arm. 

My bigger point is that there's nothing here that _requires_ a macro to run (e.g., it's not generating any code based on what's being fed in). You could just as easily have `publish(bus, event_type, data)` and `publish_with_key(bus, event_type, data, key)` functions and achieve the same effect. If macros aren't required and there's not a strong ergonomics/usability argument for them, I think it's usually better from a readability perspective to use plain functions.

Also, this macro still leaks details at the call site. Things are fine when you're calling the non-routing key version, but when you call it with a routing key, you have to wrap it in a `Some`, even though the way you call the macro suggests that there's no optionality involved. That is, why would I (as a user, not as another function) want to call 

```rust
publish_event!(bus, type, data, with_routing_key: None)
```
?

Additionally, if you just want to pass in arbitrary objects that are destined to be turned into JSON, you can specify a `Serialize` trait bound on that parameter:

```rust
fn publish<D>(bus: ..., event_type: ..., data: D) 
    where D: Serialize
{
    # do stuff to turn data into JSON
}
```

If you do that, though, at some point in the future you might want to constrain things a bit more. As written, this will take _anything_ that implements `Serialize`, but you may (for peace of mind, comprehensibility, and general sanity preservation) want to tighten that up. You could instead imagine a world where `BuilderEvent` is a trait that has `Serialize` as a _supertrait_. In that world, the `publish` function would instead have a trait bound on `BuilderEvent` (at least conceptually; I know I'm changing things up here from what you currently have written). You still retain the ability to ensure that whatever you pass in can be turned into JSON, but you also specify that you don't want just _anything_, but something that is specifically one of your events.

Reply:
I've refactored the event creation and publishing API quite a bit in  https://github.com/habitat-sh/builder/pull/1540/commits/9f5f1e41cf492cd955da60b6f382128239b3a073 

`publish()` is now a method of `BuilderEvent` which feels to me a more elegant way to both create and publish the event in one line at the calling site:

```
BuilderEvent::new(EventType::PackageChannelMotion, NoKey, &auditevent)
      .publish(&state.eventproducer).await;
```